--- a/scripts/package-metadata.pl
+++ b/scripts/package-metadata.pl
@@ -304,11 +304,11 @@
 			print "\t\t".($pkg->{tristate} ? 'tristate' : 'bool')." $title\n";
 			print "\t\tdefault y if DEFAULT_".$pkg->{name}."\n";
 			unless ($pkg->{hidden}) {
-				my @def = ("ALL");
-				if (!exists($pkg->{repository})) {
-					push @def, "ALL_NONSHARED";
-				}
+				my @def;
 				if ($pkg->{name} =~ /^kmod-/) {
 					push @def, "ALL_KMODS";
+				} else {
+					push @def, "ALL";
+					push @def, "ALL_NONSHARED" if !exists($pkg->{repository});
 				}
				$pkg->{default} ||= "m if " . join("||", @def);
			}
--- a/config/Config-build.in
+++ b/config/Config-build.in
@@ -10,4 +10,3 @@
 	config ALL_NONSHARED
 		bool "Select all target specific packages by default"
-		select ALL_KMODS
 		default BUILDBOT
@@ -18,4 +17,3 @@
 	config ALL
 		bool "Select all userspace packages by default"
-		select ALL_KMODS
 		select ALL_NONSHARED
--- a/include/kernel-defaults.mk
+++ b/include/kernel-defaults.mk
@@ -97,6 +97,8 @@
 	cmp -s $(LINUX_DIR)/.config.set $(LINUX_DIR)/.config.prev || { \
 		cp $(LINUX_DIR)/.config.set $(LINUX_DIR)/.config; \
 		cp $(LINUX_DIR)/.config.set $(LINUX_DIR)/.config.prev; \
 	}
+	$(_SINGLE)$(KERNEL_MAKE) olddefconfig
+	cp $(LINUX_DIR)/.config $(LINUX_DIR)/.config.set
 	$(_SINGLE) [ -d $(LINUX_DIR)/user_headers ] || $(KERNEL_MAKE) INSTALL_HDR_PATH=$(LINUX_DIR)/user_headers headers_install
 	grep '=[ym]' $(LINUX_DIR)/.config.set | LC_ALL=C sort | mkhash md5 > $(LINUX_DIR)/.vermagic
